src/main/java/transforms/PaddingTransform.java

```diff
@@ -0,0 +1,101 @@
+package transforms;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.openlattice.shuttle.transformations.Transformation;
+import com.openlattice.shuttle.transformations.Transformations;
+import com.openlattice.shuttle.util.Cached;
+import com.openlattice.shuttle.util.Constants;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+
+public class PaddingTransform extends Transformation<Map<String, String>> {
+    private final String pattern;
+    private final int length;
+    private final boolean pre;
+    private final boolean cutoff;
+
+    /**
+     * A transform that pads the input string to a certain length by appending or prepending a repeating pattern
+     *
+     * @param pattern:            pad will be constructed by repeating this string
+     * @param length:             target length of resulting string
+     * @param pre:                whether to prepend (as opposed to append)
+     * @param cutoff:             whether to trim the string to the target size if it's larger
+     *
+     */
+    @JsonCreator
+    public PaddingTransform(
+            @JsonProperty( Constants.PATTERN ) String pattern,
+            @JsonProperty( Constants.LENGTH ) int length,
+            @JsonProperty( Constants.PRE ) Optional<Boolean> pre,
+            @JsonProperty( Constants.CUTOFF ) Optional<Boolean> cutoff
+            ) {
+        this.pattern = pattern;
+        this.length = length;
+        this.pre = pre.orElse(false);
+        this.cutoff = cutoff.orElse(false);
+    }
+
+    @JsonProperty( Constants.PATTERN )
+    public String getPattern() {
+        return pattern;
+    }
+
+    @JsonProperty( Constants.LENGTH )
+    public int getLength() {
+        return length;
+    }
+
+    @JsonProperty( Constants.PRE )
+    public boolean getPre() {
+        return pre;
+    }
+
+    @JsonProperty( Constants.CUTOFF )
+    public boolean getCutoff() {
+        return cutoff;
+    }
+
+    @Override
+    public Object applyValue( String o ) {
+
+        if (length < 0) {
+            throw new IllegalStateException( "Negative length given." );
+        }
+
+        if (pattern.equals("")) {
+            throw new IllegalStateException( "Empty pattern given." );
+        }
+
+        if (o.length() > length) {
```

Comment:
This gets pretty complicated to read when there's this many nested if / else statements. I think this could be made more accessible by creating some helper functions to handle isolated pieces of this logic, especially ones that are repeated. i.e., you could handle the trimming logic in one function like
```
String getTrimmedString( String input ) {
  if ( !cutoff || input.length() <= length ) {
    return input;
  }

  int startIndex = pre ? 0 : input.length() - length;
  return input.substring( startIndex, length );
}
```

And then the rest of the logic gets way simpler --
```
StringBuilder builder = new StringBuilder( o );
while ( length > builder.length() ) {
  if ( pre ) {
    builder.insert( 0, pattern );
  } else {
    builder.append( pattern );
  }
}
return getTrimmedString( builder.toString() );
```

Reply:
+100

Reply:
The code you're suggesting won't give the right output if for example:
```
o = ""
pattern = "01"
length = 3
pre = false
cutoff = false
```
Should return: `010`
Would return: `0101`

I just changed it so that it uses the helper function only in the case where the input string is already too long, and does everything else the same. Also added comments for readability.

Reply:
Well `getTrimmedString` wouldn't execute until `length > builder.length()` evaluates to false. This is cleaner but it's still better to avoid duplicating the while loop code and the substring logic. 

No guarantees that the code I wrote is correct or even compiles btw, but I don't think it would return `0101` in that example (since cutoff is false, based on my understanding of what cutoff means)

Reply:
Doesn't duplicating the while loop + substring logic minimize the expected number of comparisons?

My thought process when making this transform was we need a way to pad strings to a specific length, where the pad can be any string. The `cutoff` parameter was just meant to answer the question of what to do with inputs that are too large to pad.